File: project/app/src/main/java/org/owntracks/android/data/repos/ContactsRepo.java
package org.owntracks.android.data.repos;

import androidx.lifecycle.MutableLiveData;

import org.owntracks.android.model.FusedContact;
import org.owntracks.android.model.messages.MessageCard;
import org.owntracks.android.model.messages.MessageLocation;

import java.util.Map;


public interface ContactsRepo {
    MutableLiveData<Map<String, FusedContact>> getAll();

    FusedContact getById(String id);

    void clearAll();
    void remove(String id);

    void update(String id, MessageLocation m);
    void update(String id, MessageCard m);

    long getRevision();
}

File: project/app/src/main/java/org/owntracks/android/data/repos/ContactsRepo.kt
package org.owntracks.android.data.repos

import kotlinx.coroutines.flow.SharedFlow
import org.owntracks.android.model.Contact
import org.owntracks.android.model.messages.MessageCard
import org.owntracks.android.model.messages.MessageLocation
import org.owntracks.android.model.messages.MessageTransition

/**
 * A storage for Contacts. This is populated on startup by whatever's in the upstream recorder, so we don't persist this
 * in anything other than a simple Map. On changes, we emit a [ContactsRepoChange] event.
 */
interface ContactsRepo {
    val all: Map<String, Contact>
    val repoChangedEvent: SharedFlow<ContactsRepoChange>
    fun getById(id: String): Contact?
    suspend fun clearAll()
    suspend fun remove(id: String)
    suspend fun update(id: String, messageLocation: MessageLocation)
    suspend fun update(id: String, messageTransition: MessageTransition)
    suspend fun update(id: String, messageCard: MessageCard)
}

sealed class ContactsRepoChange {
    data class ContactAdded(val contact: Contact) : ContactsRepoChange()
    data class ContactRemoved(val contact: Contact) : ContactsRepoChange()
    data class ContactLocationUpdated(val contact: Contact) : ContactsRepoChange()
    data class ContactCardUpdated(val contact: Contact) : ContactsRepoChange()
    data object AllCleared : ContactsRepoChange()
}
